pinot-controller/src/main/java/org/apache/pinot/controller/util/BrokerServiceHelper.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller.util;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.pinot.controller.ControllerConf;
+import org.apache.pinot.controller.helix.core.PinotHelixResourceManager;
+import org.apache.pinot.core.routing.TimeBoundaryInfo;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.apache.pinot.spi.utils.JsonUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Helper class for interacting with broker APIs.
+ */
+public class BrokerServiceHelper {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BrokerServiceHelper.class);
+  private static final String TIME_BOUNDARY_INFO_API_PATH = "/debug/timeBoundary/%s";
+  private static final int DEFAULT_REQUEST_TIMEOUT_MS = 30_000;
+  private final PinotHelixResourceManager _pinotHelixResourceManager;
+  private final ControllerConf _controllerConf;
+  private final Executor _executorService;
+  private final PoolingHttpClientConnectionManager _connectionManager;
+  private CompletionServiceHelper _completionServiceHelper;
```

Comment:
Is this safe to be cached and shared? I think we need to recompute the endpoint map every time

Reply:
The `CompletionServiceHelper` is not initialized by default and is constructed everytime. A set() method is provided purely for testing purposes which makes writing tests easier.

Reply:
On line 95, we are reusing the cached one right?

Reply:
I see your point. Let me fix this.

Reply:
I'm using a method-local version of the `CompletionServiceHelper` and the same is not re-used between subsequent calls. The instance specific `CompletionServiceHelper` should be set only for testing purposes. I have added comment in the set method regarding this.